rfd/0207-encrypted-session-recordings.md

```diff
@@ -0,0 +1,284 @@
+[here](https://age-encryption.org/v1). Officially supported key algorithms are
+limited to X25519 (recommended by the spec), Ed25519, and RSA. Support for other
+algorithms would either have to be requested from the upstream or manually
+implemented as a custom plugin.
+
+### Config Changes
+
+Enabling encrypted session recordings requires configuring one or more public key
+recipients that will be used during encryption. All private key identities
+associated with configured recipients will be able to decrypt the resulting files.
+This configuration is possible either through file configuration or dynamically
+though the `session_recording_config` resource.
+
+```yaml
+# teleport.yaml
+version: v3
+auth_service:
+  session_recording_recipients:
+  - recipient_type: age_x25519
+    public_key: age18l5tzttm32yqvh9pw3dlwp8h4qlvqnx8a40hf7svyag8hnumlsus43gc0y
+```
+
+```yaml
+# session_recording_config resource
+kind: session_recording_config
+version: v2
+spec:
+  recipients:
+  - recipient_type: age_x25519
+    public_key: age18l5tzttm32yqvh9pw3dlwp8h4qlvqnx8a40hf7svyag8hnumlsus43gc0y
+```
+
+### Proto Specification
+```proto
+message SessionRecordingConfigSpecV2 {
+  // (existing fields omitted)
+
+  // Encrypted file recipients for recorded sessions.
+  // If non-empty then all session recordings are encrypted end-to-end
+  // (filesystem to end storage).
+  repeated EncryptedFileRecipient recipients = 3 [(gogoproto.jsontag) = "recipients"];
+}
+
+// Illustrative, in practice a string is used for config readability.
+enum EncryptedFileRecipientType {
+  ENCRYPTED_FILE_RECIPIENT_TYPE_UNSPECIFIED = 0;
+
+  // Age X25519 recipient string. Either the recipient alone or concatenated with a wrapped
+  // identity.
+  // * https://github.com/C2SP/C2SP/blob/8991f70ddf8a11de3a68d5a081e7be27e59d87c8/age.md#the-x25519-recipient-type
+  ENCRYPTED_FILE_RECIPIENT_TYPE_AGE_X25519 = 1;
+  // Ed25519 public key marshaled as an ssh-ed25519 public key.
+  ENCRYPTED_FILE_RECIPIENT_TYPE_SSH_ED25519 = 2;
+  // RSA public key marshaled as an ssh-rsa public key.
+  // RSA keys must have at least 2048 bits.
+  ENCRYPTED_FILE_RECIPIENT_TYPE_SSH_RSA = 3;
+  // Ed25519 public key marshaled in PKIX PEM form.
+  ENCRYPTED_FILE_RECIPIENT_TYPE_ED25519 = 4;
+  // RSA public key marshaled in PKIX PEM form.
+  // RSA keys must have at least 2048 bits.
+  ENCRYPTED_FILE_RECIPIENT_TYPE_RSA = 5;
+}
+
+message EncryptedFileRecipient {
+  // Recipient public key type.
+  // * age_x25519
+  // * ssh_ed25519 
+  // * ssh_rsa
+  // * ed25519
+  // * rsa
+  string recipient_type = 1 [(gogoproto.jsontag) = "recipient_type"];
+  // Public key marshaled according to recipient_type.
+  string public_key = 2 [(gogoproto.jsontag) = "public_key"];
+}
+```
+
+### Key Generation
+
+Admins are free to configure their own public key recipients provided they are
+generated by a supported algorithm.
+
+As a convenience, we will also add a new `keygen` utility to `tsh` for key
```

Comment:
Would this be more appropriate for `tctl` than `tsh`? It feels like an administrative workflow to me, not an end-user workflow.

Reply:
Yeah I think that makes more sense. At one point it was a `tctl` command, but the similarity to `ssh-keygen` and the fact that `tsh` _has_ to integrate with HSMs in order to playback made me wonder if it would be worth it to just do the entire workflow with `tsh`. `Administrative workflows running through `tctl` seems like a more compelling reason, though